distributed/comm/zmq.py

```diff
@@ -0,0 +1,218 @@
+from __future__ import print_function, division, absolute_import
+
+from datetime import timedelta
+import logging
+import struct
+import sys
+
+from tornado import gen, ioloop
+
+import zmq
+from zmq.eventloop import ioloop as zmqioloop
+from zmq.eventloop.future import Context
+
+from .. import config
+from .core import connectors, listeners, Comm, CommClosedError, Listener
+from .utils import to_frames, from_frames, parse_host_port, unparse_host_port
+from . import zmqimpl
+
+
+logger = logging.getLogger(__name__)
+
+
+def install():
+    ioloop.IOLoop.clear_current()
+    ioloop.IOLoop.clear_instance()
+    zmqioloop.install()
+
+    inst = ioloop.IOLoop.instance()
+    if not isinstance(inst, zmqioloop.IOLoop):
+        raise RuntimeError("failed to install ZeroMQ IO Loop, got %r" % (inst,))
+
+install()
+
+
+
+NOCOPY_THRESHOLD = 1000 ** 2   # 1 MB
+
+
+#async_ctx = Context()
+## Workaround https://github.com/zeromq/pyzmq/issues/962
+#async_ctx.io_loop = None
+
+#def make_socket(sockty):
+    #sock = async_ctx.socket(sockty)
+    #return sock
+
+ctx = zmqimpl.Context()
+
+def make_socket(sockty):
+    sock = ctx.socket(sockty)
+    return sock
+
+
+def set_socket_options(sock):
+    """
+    Set common options on a ZeroMQ socket.
+    """
+    # XXX set context default options instead?
+    sock.set(zmq.RECONNECT_IVL, -1)  # disable reconnections
+    sock.set(zmq.IPV6, True)
```

Comment:
I'm hoping this way ZeroMQ might actually inform us of disconnects. Though at this point I'm considering the ZeroMQ transport a mere experiment, as ZeroMQ's characteristics doesn't make it very friendly to work with.